OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala

```diff
@@ -0,0 +1,1430 @@
+                val defIndex = p.asVar.definedBy.head
+                if (defIndex < 0) {
+                    return None
+                }
+                val defSiteStm = tacCode.stmts(defIndex)
+                if (defSiteStm.isAssignment) {
+                    val call = defSiteStm.asAssignment.expr.asFunctionCall
+                    if (!intentMethods.contains(call.name)) {
+                        return Some[(ObjectType, String)](call.declaringClass.asObjectType -> call.name)
+                    }
+                    return searchParamsStatic(call, tacCode)
+                }
+            }
+        }
+        Some[(ObjectType, String)](functionCall.declaringClass.asObjectType -> functionCall.name)
+    }
+
+    /**
+     * searches the parameters of functionCall to find the origin of the intent and return the class and method
+     * that returns the searched intent
+     *
+     */
+    def searchParamsVirtual(
+                             functionCall: MethodCall[DUVar[ValueInformation]],
+                             tacCode: AITACode[TACMethodParameter, ValueInformation]
+                           ): Option[(ObjectType, String)] = {
+        functionCall.params.foreach { p ⇒
+            if (p.asVar.value.toString.startsWith(intentStart.head) || p.asVar.value.toString.startsWith(intentStart(1))) {
+
+                val defIndex = p.asVar.definedBy.head
+                if (defIndex < 0) {
+                    return None
+                }
+                val defSiteStm = tacCode.stmts(defIndex)
+                if (defSiteStm.isAssignment && (defSiteStm.asAssignment.expr.isVirtualFunctionCall ||
+                  defSiteStm.asAssignment.expr.isStaticFunctionCall)) {
+
+                    val call = defSiteStm.asAssignment.expr.asFunctionCall
+                    if (!intentMethods.contains(call.name)) {
+                        return Some[(ObjectType, String)](call.declaringClass.asObjectType -> call.name)
+                    }
+                    return searchParamsStatic(call, tacCode)
+                }
+            }
+        }
+        Some[(ObjectType, String)](functionCall.declaringClass.asObjectType -> functionCall.name)
+    }
+
+    def getComponentType(
+                          name: String
+                        ): String = {
+        if (activityStartMethods.contains(name)) {
+            return activity
+        }
+        if (serviceMethods.contains(name)) {
+            return service
+        }
+        if (broadcastReceiverStartMethods.contains(name)) {
+            return receiver
+        }
+        undef
+    }
+
+    def reconstructIntent(
+                           c:           NonVirtualMethodCall[V],
+                           tacCode:     AITACode[TACMethodParameter, ValueInformation],
+                           useSites:    List[Int],
+                           m:           Method
+    ): Unit = {
+        val p = c.params
+
+        p.size match {
+            case 0 ⇒ //Intent()
```

Comment:
Can't you match the actual signature of the call?

Reply:
Yes this may also provide more accurate results. Can I use the method descriptor for this?

Reply:
Yes, c.descriptor gives you the signature (a.k.a. descriptor) of the method called.